dbclient/ScimClient.py

```diff
@@ -353,6 +411,37 @@ def assign_user_roles(self, current_user_ids, error_logger, user_log_file='users
                     update_resp = self.patch(f'/preview/scim/v2/Users/{user_id}', patch_roles)
                     logging_utils.log_response_error(error_logger, update_resp)
 
+    def assign_service_principal_roles(self, current_service_principal_ids, error_logger, service_principal_log_file='service_principals.log'):
+        """
+        assign service principal roles that are missing after adding group assignment
+        Note: There is a limitation in the exposed API. If a service principal is assigned a role permission & the permission
+        is granted via a group, we can't distinguish the difference. Only group assignment will be migrated.
```

Comment:
Not sure I understood this part

Reply:
It's copy pasted from the same steps done with the users. I didn't fully understand the API limitation thing either, but I copied it to play it safe.
What the user role function does is a diff of the current user roles attribute and the new one, and PATCH the diff. I guess this is about the fact that the "roles" attr might be populated from permissions inherited from groups as well as direct ones and they cannot be differentiated... but the final sentence is still a big question mark for me.